repos/system_upgrade/el7toel8/actors/checkpcidrivers/libraries/checkpcidrivers.py

```diff
@@ -0,0 +1,164 @@
+from leapp.models import PCIDevices, RestrictedPCIDevices
+from leapp.reporting import create_report
+
+
+def render_report(
+    restricted_driver_names_on_host,
+    restricted_pci_ids_on_host,
+    restricted_devices_drivers,
+    restricted_devices_pcis,
+    inhibit_upgrade=False,
+):
+    """
+    Render the report for restricted PCI devices in use.
+
+    :param inhibit_upgrade: if True, the report will have Inhibitor flag and
+        the upgrde will be inhibited.
+    """
+    # generate strings for the report with the restricted drivers
+    unavailable_driver_names = ",\n".join(
+        driver
+        for driver in restricted_driver_names_on_host
+        if restricted_devices_drivers[driver].available_rhel8 == 0
+    )
+    unavailable_pci_ids = ",\n".join(
+        pci_id
+        for pci_id in restricted_pci_ids_on_host
+        if restricted_devices_pcis[pci_id].available_rhel8 == 0
+    )
+    unsupported_driver_names = ",\n".join(
+        driver
+        for driver in restricted_driver_names_on_host
+        if restricted_devices_drivers[driver].supported_rhel8 == 0
+    )
+    unsupported_pci_ids = ",\n".join(
+        pci_id
+        for pci_id in restricted_pci_ids_on_host
+        if restricted_devices_pcis[pci_id].supported_rhel8 == 0
+    )
+
+    # Render the report
+    title = "Detected PCI drivers that are restricted on RHEL8."
+    summary = (
+        "There are following drivers that are restricted on RHEL 8 system.\n\n"
+        "Devices which are unavailable on RHEL8:\n "
+        "{} (driver names),\n{} (pci ids)\n\n"
+        "Devices which are unsupported on RHEL8:\n "
+        "{} (driver names),\n{} (pci ids)\n\n"
+    ).format(
+        unavailable_driver_names,
+        unavailable_pci_ids,
+        unsupported_driver_names,
+        unsupported_pci_ids,
+    )
+    reports = [
+        reporting.Title(title),
+        reporting.Summary(summary),
+        reporting.Severity(reporting.Severity.HIGH),
+        reporting.Tags([reporting.Tags.KERNEL, reporting.Tags.DRIVERS]),
+    ]
+    if inhibit_upgrade:
+        reports += [reporting.Flags([reporting.Flags.INHIBITOR])]
+    return reports
+
+
+def create_dict_lookup(driver_names, key):
+    """
+    Creates a dict lookup from the list of Models.
+
+    :param driver_names: List[leapp.models.Model[K, V]]
+    :param key: a key value, that will be used as a primary key to generate a
+        result Dict
+    :return: Dict[leapp.models.Model[K,V].key, leapp.models.Model[K,V]]
+    :raises ValueError: if there are duplicates of leapp.models.Model[K,V].key
+    :raises AttributeError: if key not in K (Model attributes)
+
+    Example:
+    >>> create_dict_lookup([Model(a=1, b=2), Model(a=2, b=3)], key="b") == {
+    >>>     2: Model(a=1, b=2),
+    >>>     3: Model(a=2, b=3),
+    >>> }
+    """
+    lookup = {getattr(item, key): item for item in driver_names}
+    if len(lookup) != len(driver_names):
+        raise ValueError()
```

Comment:
I was thinking about this and I didn't find it possible, because:
1. The implementation in the rpms returns a set. I need a dict
2. The rpms implementation assumes to generate the lookup from the whole Model, but I need lookups from Model attributes
3. I don't think we have enough identical usecases yet to move the functionality to the common lib